src/main/scala/com/databricks/labs/overwatch/pipeline/Pipeline.scala

```diff
@@ -118,76 +118,128 @@ class Pipeline(
       logger.log(Level.INFO, msg)
       if (config.debugFlag) println(msg)
 
-      // All active records with new DBU pricing
-      val newDBUPriceRecords = lastRunPricing
-        .filter('rnk === 1 && 'activeUntil.isNull)
-        .withColumn("interactiveDBUPrice", lit(config.contractInteractiveDBUPrice))
-        .withColumn("automatedDBUPrice", lit(config.contractAutomatedDBUPrice))
-        .withColumn("sqlComputeDBUPrice", lit(config.contractSQLComputeDBUPrice))
-        .withColumn("jobsLightDBUPrice", lit(config.contractJobsLightDBUPrice))
+      // configured costs by sku as a column
+      val activeCost = when('sku === "interactive", config.contractInteractiveDBUPrice)
+        .when('sku === "automated", config.contractAutomatedDBUPrice)
+        .when('sku === "sqlCompute", config.contractSQLComputeDBUPrice)
+        .when('sku === "jobsLight", config.contractJobsLightDBUPrice)
+        .otherwise(lit(0.0))
+
+      // records that were already expired
+      val originalExpiredRecords = lastRunDBUCosts.filter(!activeRecord)
+
+      // active costs with no changes
+      val unchangedActiveCosts = lastRunDBUCosts.filter(activeRecord && 'contract_price === activeCost)
+
+      // newly active records due to costing change
+      val newDBUPriceRecords = lastRunDBUCosts
+        .filter(activeRecord && 'contract_price =!= activeCost)
+        .withColumn("contract_price", activeCost)
         .withColumn("activeFrom", lit(pipelineSnapTime.asDTString).cast("date"))
         .withColumn("activeUntil", lit(null).cast("date"))
+        .withColumn("isActive", lit(true))
         .withColumn("Pipeline_SnapTS", lit(pipelineSnapTime.asColumnTS))
         .withColumn("Overwatch_RunID", lit(config.runID))
 
-      // All originally active records with expiration date of yesterday
-      val expiredRecords = lastRunPricing
-        .filter('rnk === 1)
-        .withColumn("activeUntil",
-          when('activeUntil.isNull, date_sub(lit(pipelineSnapTime.asDTString).cast("date"), 1))
-            .otherwise('activeUntil)
-        )
-
-      // New pricing records ++ expired pricing records ++ historical records
-      val newCloudDetailsDF = newDBUPriceRecords
-        .unionByName(expiredRecords)
-        .unionByName(lastRunPricing.filter('rnk > 1))
+      // active costs to be expired due to replacement
+      val expiringRecords = lastRunDBUCosts
+        .filter(activeRecord&& 'contract_price =!= activeCost)
+        .withColumn("activeUntil", lit(pipelineSnapTime.asDTString).cast("date"))
+        .withColumn("isActive", lit(false))
+
+      // historical records ++ new pricing records ++ newly expired pricing records ++ unchanged active records
+      val updatedDBUCostDetail = newDBUPriceRecords
+        .unionByName(expiringRecords)
+        .unionByName(unchangedActiveCosts)
+        .unionByName(originalExpiredRecords)
         .drop("rnk")
-        .repartition(4)
+        .coalesce(1)
```

Comment:
Why 1? If to minimize number of files, I would rather use optimized writes

Reply:
because this lookup actually only holds about 4 records (maybe a few more over time but always less than 100...it's just sooo tiny.